packages/api-server/src/error.ts

```diff
@@ -0,0 +1,160 @@
+//  Error code from JSON-RPC 2.0 spec
+//  reference: http://www.jsonrpc.org/specification#error_object
+
+import { JSONRPCError } from "jayson";
+import { COMPATIBLE_DOCS_URL } from "./methods/constant";
+
+export interface ErrorKind {
+  code: number;
+  message: string;
+}
+
+export class AppError extends Error implements JSONRPCError {
+  code: number;
+  data?: object;
+
+  constructor(kind: ErrorKind, data?: object) {
+    super(kind.message);
+    this.name = kind.message;
+    this.code = kind.code;
+    this.data = data;
+  }
+
+  addContext(context: Object): AppError {
```

Comment:
the original padContext has the ability to provide the params call chain in `validator.ts` like `callObj -> gasLimit -> hexNumber without 0x prefix`, can we keep this feature?

Reply:
I didn't realize that we need that params call chain. Instead, `addContext` only insert new items into `error.data`. I paste the `addContext` usages below:

```shell
git:(main)  gh pr checkout 352

git:(apply-app-error)  grep -r 'addContext' packages/api-server/src
packages/api-server/src/methods/validator.ts:226:    return err.addContext({ type: "address" });
packages/api-server/src/methods/validator.ts:249:    return err.addContext({ type: "blockHash" });
packages/api-server/src/methods/validator.ts:269:    return err.addContext({ type: "txHash" });
packages/api-server/src/methods/validator.ts:300:    return err.addContext({ type: "BlockParameter::Number" });
packages/api-server/src/methods/validator.ts:329:      return toErr.addContext({ property: "to" });
packages/api-server/src/methods/validator.ts:337:      return fromErr.addContext({ property: "from" });
packages/api-server/src/methods/validator.ts:345:      return gasErr.addContext({ property: "gasPrice" });
packages/api-server/src/methods/validator.ts:353:      return gasLimitErr.addContext({ property: "gasLimit" });
packages/api-server/src/methods/validator.ts:361:      return valueErr.addContext({ property: "value" });
packages/api-server/src/methods/validator.ts:369:      return dataErr.addContext({ property: "data" });
packages/api-server/src/methods/validator.ts:426:    return err.addContext({ type: "topic" });
packages/api-server/src/methods/validator.ts:455:        return err.addContext({ type: "topicString[]" });
packages/api-server/src/methods/validator.ts:484:      return fromBlockErr.addContext({ property: "fromBlock" });
packages/api-server/src/methods/validator.ts:492:      return toBlockErr.addContext({ property: "toBlock" });
packages/api-server/src/methods/validator.ts:502:          return addressErr.addContext({
packages/api-server/src/methods/validator.ts:511:        return addressErr.addContext({ property: "address", type: "address" });
packages/api-server/src/methods/validator.ts:529:        return topicErr.addContext({
```

Reply:
In my opinion, I think a properly designed context could alternate the call chain.

Reply:
Mostly, An error entry should contain the below information I expect:
- RPC method
- Error message
- The index of invalid parameter, or property name of invalid structure
- The declared type of that parameter

Reply:
> In my opinion, I think a properly designed context could alternate the call chain.

I agree, the main idea here is to help locate the illegal param position in the validator.